### PR TITLE
Increased String::num default decimal precision

### LINKED ISSUE #34541: to_json loses float precision
**Godot version:** 3.1.2.stable.official

**OS/device including version:** Linux (Ubuntu 18.04)

**Issue description:**

Storing floating point values with `to_json` loses precision.

I would have expected `to_json` to always store full information, so that a serialization/deserialization round-trip is guaranteed to be lossless. If lossy representation is desired for size reduction, it could be passed as a flag `lossy_flots` like `to_json(data, true)`.

Note: The problem applies to recursive application of `to_json` as well, which makes manual work-arounds on user side difficult, e.g. in a situation where a high level object is to be serialized, but a deeply nested float must be stored in full precision.

**Steps to reproduce:**

```gdscript
# To check actual precision of floats:
print("%.16f" % PI)
print("%.16f" % (1.0 / 3.0))

# Results after going through `to_json`
print(to_json(PI))
print(to_json(1.0 / 3.0))

# Round-trip test
print("%.16f" % (PI - parse_json(to_json(PI))))
```
Outputs:
```
3.1415926535897931
0.3333333333333333
3.141593
0.333333
-0.0000003464102067
```

**Minimal reproduction project:** Let me know if needed...

### PR DESCRIPTION
**Pull request by simojoki**
Fixes #34541

Increased MAX_DIGITS to 16, and made String::num use
MAX_DIGITS consistently. If -1 is passed as
decimal precision to String::num, it now gets changed to
MAX_DIGITS instead of using printf default(which is 6)

Also made the types used in floating-point math more
consistent in a few places.

Code from the original issue works correctly with these changes:

```
# To check actual precision of floats:
print("%.16f" % PI)
print("%.16f" % (1.0 / 3.0))

# Results after going through `to_json`
print(to_json(PI))
print(to_json(1.0 / 3.0))

# Round-trip test
print("%.16f" % (PI - parse_json(to_json(PI))))
```

Now outputs:
```
3.1415926535897931
0.3333333333333333
3.1415926535897931
0.3333333333333333
0.0000000000000000
```

### CODE REVIEW

**Review by aaronfranke — CHANGES_REQUESTED** [review]
This results in false digits being printed, since double does not have this much precision (or for that matter, a fixed amount of precision at all). For example, in both master and your PR:

	print("%.16f" % 3.1415926535897931) # 3.1415926535897931
	print("%.16f" % 3.1415926535897932) # 3.1415926535897931
	print("%.16f" % 3.1415926535897933) # 3.1415926535897931
	print("%.16f" % 3.1415926535897934) # 3.1415926535897936
	print("%.16f" % 3.1415926535897935) # 3.1415926535897936
	print("%.16f" % 3.1415926535897936) # 3.1415926535897936

A better idea would be to calculate the amount of decimal digits depending on the magnitude of the number. On top of your PR, at line 1036 I changed the code to this:

	if (p_decimals < 0) {
		p_decimals = 14 - (int)floor(log10(p_num));
	}
	if (p_decimals > MAX_DIGITS) {
		p_decimals = MAX_DIGITS;
	}

Test code with correct output:

	print("%.16f" % PI) # 3.1415926535897931, forcing 16 digits still works
	print(to_json(PI))  # 3.14159265358979
	print(PI)           # 3.14159265358979
	print(PI * 10)      # 31.4159265358979
	print(PI * 100)     # 314.159265358979
	print(PI * 1000)    # 3141.59265358979

**Comment by Heikki00** [comment]
@aaronfranke Good points, added the changes.

File: core/ustring.cpp [hunk 1]
```diff
@@ -1267,7 +1265,7 @@ String String::num_real(double p_num) {
 
 	if ((int)p_num != p_num) {
 
-		double dec = p_num - (float)((int)p_num);
+		double dec = p_num - (double)((int)p_num);
 
 		int digit = 0;
 		int decimals = MAX_DIGITS;
```
Comment (aaronfranke) [inline core/ustring.cpp]: Since this method is called `num_real`, I think it should reflect the precision of `real_t` - the values should be kept as double, but we should change this to use less digits (the previous value of `MAX_DIGITS`) unless `REAL_T_IS_DOUBLE`.

```
#if REAL_T_IS_DOUBLE
		int decimals = MAX_DIGITS;
#else
		int decimals = 6;
#endif
```

**Review by aaronfranke — COMMENTED** [review]
This is now the output for the example code in the OP (all of these values are correct and expected):

```
3.1415926535897931
0.3333333333333333
3.14159265358979
0.333333333333333
0.0000000000000031
```

The non-zero value for the last line is expected, since it is only the difference of having the false digits (because they were force-printed) vs not having them. If I adapt `is_zero_approx` to use the `1e-14` epsilon appropriate for double precision floats, it outputs true:

    print(abs(PI - parse_json(to_json(PI))) < 1e-14) # True

**Comment by Heikki00** [comment]
@aaronfranke Added the changes to String::num_real. While I was at it, I renamed MAX_DIGITS to MAX_DECIMALS since it doesn't really affect the number of total digits. I also increased it's value to 32 since a lot more than 16 decimals can be needed at times, and no function defaults to MAX_DECIMALS so there isn't many drawbacks to it being a larger number.

**Review by aaronfranke — APPROVED** [review]
All of these logic changes should be good, as well as the renaming of MAX_DIGITS to MAX_DECIMALS. I don't think this should be cherrypicked.

The MAX_DECIMALS value being larger than 16 is only useful for very small values or for debugging, so I don't think it's useful very often, but then again it only comes into effect when the user manually specifies they want so many digits, so I think it's a good change.

The title is a bit inaccurate now, what this PR does is more like "Increased maximum precision of number to String functions and improve auto-detection of the number of digits based on the order of magnitude of the number", but that's a bit large for the title, so feel free to keep it as-is.

**Comment by bluenote10** [comment]
Shouldn't we rather rely on `%g` as recommended by [this discussion](https://stackoverflow.com/a/21162120/1804173)?

How would the current approach handle for instance:

```gdscript
print(to_json(1.175494e-38))  # should avoid becoming zero
print(to_json(3.4028237e38))  # should avoid large number of trailing zeros
```

**Comment by aaronfranke** [comment]
@Heikki00 When you get a chance, this needs to be rebased on the latest master branch.

**Comment by bluenote10** [comment]
@aaronfranke Do you think this solves #34541? Unless I'm misreading the code, values with negative exponents can still be lossy, and using `%f` for large positive exponents can lead to a huge json output sizes compared to exponential notation.

The perfect solution would be [ryu](https://github.com/ulfjack/ryu), but that is beyond the scope I guess. The most sensible (fast) `sprintf` based solution to my knowledge is to use `"%.17g"`, which always roundtrips perfectly, but has the downside that some numbers like 0.1 get represented with "too much precision" (0.10000000000000001). This leads to a considerable size overhead in the json, but in any case the overhead is much smaller than using `%f` e.g. because `1.7976931348623157e+308` becomes 
```
179769313486231570000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
```
which typically also slows down parsing.

**Comment by aaronfranke** [comment]
> but has the downside that some numbers like 0.1 get represented with "too much precision" (0.10000000000000001).

This is what's called false digits. In many cases these methods are used to display numbers to the user. Ideally, we should avoid printing digits which can't reliably store values when going decimal -> binary -> decimal.

There are certainly some cases where you want to guarantee binary -> decimal -> binary, in which case you can specify `"%.17g"` or similar (perhaps `to_json` needs an override for this, but I think that can be a separate PR).

I'm not concerned about `1.7976931348623157e+308` or similar values, it's an edge case. Most floats people actually use aren't nearly that high.

I posted above that the output of this test code:

```
print(to_json(PI))
print(to_json(1.0 / 3.0))
```

is this, which is the correct number of digits for decimal -> binary -> decimal:

```
3.14159265358979
0.333333333333333
```

**Review by aaronfranke — APPROVED** [review]
@akien-mga I rebased this and tested myself, should be good now.

Slightly tweaked so that it completely solves #34541 and doesn't lose any precision for JSON.

**Review by aaronfranke — APPROVED** [review]

**Review by aaronfranke — APPROVED** [review]
```gdscript
func _ready():
	# Actual value of Pi is  3.14159265358979323...
	print("%.16f" % PI)    # 3.1415926535897931, forcing 16 digits still works
	print(JSON.print(PI))  # 3.14159265358979312, for JSON we print unreliable digits to guarantee exact decoding.
	print(PI)              # 3.14159265358979
	print(PI * 10)         # 31.4159265358979
	print(PI * 100)        # 314.159265358979
	print(PI * 1000)       # 3141.59265358979
	print(0.1)             # 0.1
	print(JSON.print(0.1)) # 0.100000000000000006, for JSON we print unreliable digits to guarantee exact decoding.
```

**Review by aaronfranke — APPROVED** [review]

**Review by fire — APPROVED** [review]
Looks agreeable. 

I had problems with NaNs and infinities in floats strings but that isn't related to this pr.

**Comment by reduz** [comment]
I don' t think this is really useful and it makes numbers harder to read for the most common cases.
I suggest add an option to the code that writes JSON to specify the numerical precision instead. In 4.0 its a separate class.

**Comment by fire** [comment]
Reduz recommended I directly modify the json class for gltf2 export problems around Nan, Inf and incorrect decimals.

**Comment by aaronfranke** [comment]
@reduz I just pushed a change, what do you think? Now it prints reliable digits by default, but there is an option to print unreliable ones too. Here's some updated test code and output:

```gdscript
func _ready():
	# Actual value of Pi is  3.14159265358979323...
	print("%.16f" % PI)    # 3.1415926535897931, forcing 16 digits still works
	print(JSON.print(PI))  # 3.14159265358979
	print(JSON.print(PI, "", false, true))  # 3.14159265358979312, for JSON we can print unreliable digits to guarantee exact decoding.
	print(PI)              # 3.14159265358979
	print(PI * 10)         # 31.4159265358979
	print(PI * 100)        # 314.159265358979
	print(PI * 1000)       # 3141.59265358979
	print(0.1)             # 0.1
	print(JSON.print(0.1)) # 0.1
	print(JSON.print(0.1, "", false, true)) # 0.100000000000000006, for JSON we can print unreliable digits to guarantee exact decoding.
```

**Review by akien-mga — APPROVED** [review]
